File: static/app/views/insights/pages/agents/utils/aiMessageNormalizer.ts
import {tryParsePythonDict} from 'sentry/views/performance/newTraceDetails/traceDrawer/details/span/eapSections/aiContentDetection';
import {
  parseJsonWithFix,
  tryParseJsonRecursive,
} from 'sentry/views/performance/newTraceDetails/traceDrawer/details/utils';

export interface AIMessage {
  content: unknown;
  role: string;
}

interface NormalizedResult {
  fixedInvalidJson: boolean;
  messages: AIMessage[] | null;
}

interface AIOutputResult {
  fixedInvalidJson: boolean;
  responseObject: string | null;
  responseText: string | null;
  toolCalls: string | null;
}

type RawMessage = {
  content?: unknown;
  parts?: unknown[];
  role?: string;
  roleExplicit?: boolean;
};

type UnknownRecord = Record<string, unknown>;

type PartBuckets = {
  hasRenderableTextPart: boolean;
  objectParts: unknown[];
  textParts: string[];
  toolCalls: unknown[];
  toolResponses: UnknownRecord[];
};

// Keep this parser mirrored with src/sentry/utils/ai_message_normalizer.py.
// AI SDKs emit inconsistent shapes and their specs keep changing, so update both
// parsers together whenever adding or changing a supported format.

/**
 * Normalizes AI attribute values into a list of messages.
 *
 * Accepts every shape the codebase supports on supported attributes:
 * - parts-format array:     [{role, parts: [{type, text|content}, ...]}]
 * - content-format array:   [{role, content}]
 * - messages wrapper:       {messages: [...] | "[...]"}, optional `system`
 * - legacy {system, prompt}
 * - legacy {system, messages}
 * - plain string
 *
 * @param raw         Stringified attribute value.
 * @param defaultRole Role assigned when a message lacks one.
 */
export function normalizeToMessages(
  raw: string,
  {defaultRole}: {defaultRole: string}
): NormalizedResult {
  const {fixedInvalidJson, messages: rawMessages} = rawMessagesFromAttribute(
    raw,
    defaultRole
  );
  const messages = normalizeRawMessages(rawMessages, defaultRole);

  return {
    fixedInvalidJson,
    messages: messages.length > 0 ? messages : null,
  };
}

/**
 * Extracts assistant output from AI attribute values, splitting it into
 * response text, structured objects, and tool calls.
 *
 * Role selection rule:
 * - If at least one message declares a role, select all with `role === 'assistant'`.
 * - Otherwise, take the last message.
 */
export function extractAssistantOutput(
  raw: string,
  {defaultRole}: {defaultRole: string}
): AIOutputResult {
  const {fixedInvalidJson, messages: rawMessages} = rawMessagesFromAttribute(
    raw,
    defaultRole
  );

  if (rawMessages.length === 0) {
    return emptyOutput(fixedInvalidJson);
  }

  const selected = selectAssistantMessages(rawMessages);
  return outputFromMessages(selected, fixedInvalidJson);
}

function emptyOutput(fixedInvalidJson: boolean): AIOutputResult {
  return {
    fixedInvalidJson,
    responseText: null,
    responseObject: null,
    toolCalls: null,
  };
}

function normalizeRawMessages(
  rawMessages: RawMessage[],
  defaultRole: string
): AIMessage[] {
  const normalized: AIMessage[] = [];
  for (const msg of rawMessages) {
    const role = msg.role ?? defaultRole;
    const content = resolveMessageContent(msg, role);
    if (content === undefined || content === null || content === '') {
      continue;
    }
    normalized.push({role, content});
  }
  return normalized;
}

function outputFromMessages(
  messages: RawMessage[],
  fixedInvalidJson: boolean
): AIOutputResult {
  const textParts: string[] = [];
  const toolCallParts: unknown[] = [];
  const objectParts: unknown[] = [];
  for (const msg of messages) {
    appendOutputFromMessage(msg, {textParts, toolCallParts, objectParts});
  }

  return {
    fixedInvalidJson,
    responseText: textParts.length > 0 ? textParts.join('\n') : null,
    responseObject:
      objectParts.length > 0
        ? JSON.stringify(objectParts.length === 1 ? objectParts[0] : objectParts)
        : null,
    toolCalls: toolCallParts.length > 0 ? JSON.stringify(toolCallParts) : null,
  };
}

function rawMessagesFromAttribute(
  raw: string,
  defaultRole: string
): {fixedInvalidJson: boolean; messages: RawMessage[]} {
  const {fixedInvalidJson, parsed} = parseAttribute(raw);
  if (parsed === undefined) {
    return {fixedInvalidJson, messages: []};
  }

  return {
    fixedInvalidJson,
    messages: rawMessagesFromValue(parsed, defaultRole),
  };
}

function parseAttribute(raw: string): {fixedInvalidJson: boolean; parsed: unknown} {
  if (!looksLikeJson(raw)) {
    return {fixedInvalidJson: false, parsed: raw};
  }
  const {parsed, fixedInvalidJson}: {fixedInvalidJson: boolean; parsed: unknown} =
    parseJsonWithFix(raw);
  if (parsed !== null) {
    return {fixedInvalidJson, parsed};
  }

  const pythonParsed = tryParsePythonDict(raw);
  if (pythonParsed !== null) {
    return {fixedInvalidJson: false, parsed: pythonParsed};
  }

  return {fixedInvalidJson, parsed: undefined};
}

/**
 * Maps an already-parsed value onto a list of raw messages.
 */
function rawMessagesFromValue(value: unknown, defaultRole: string): RawMessage[] {
  if (typeof value === 'string') {
    return value.trim() ? [{role: defaultRole, content: value}] : [];
  }

  if (Array.isArray(value)) {
    return collectRawMessages(value, defaultRole);
  }

  if (!isRecord(value)) {
    return [];
  }

  if ('messages' in value) {
    return unwrapMessagesField(value, defaultRole);
  }

  if ('prompt' in value) {
    return unwrapSystemPrompt(value);
  }

  const single = toRawMessage(value, defaultRole);
  return single ? [single] : [];
}

function collectRawMessages(items: unknown[], defaultRole: string): RawMessage[] {
  const out: RawMessage[] = [];
  for (const item of items) {
    const msg = toRawMessage(item, defaultRole);
    if (msg) {
      out.push(msg);
    }
  }
  return out;
}

/**
 * OpenRouter-style wrapper: `{messages: [...] | "[...]"}`.
 * Also supports an optional `system` prepended.
 */
function unwrapMessagesField(value: UnknownRecord, defaultRole: string): RawMessage[] {
  const result: RawMessage[] = [];
  const system = value.system;
  if (system) {
    result.push({role: 'system', roleExplicit: true, content: system});
  }

  const inner = value.messages;
  if (Array.isArray(inner)) {
    result.push(...collectRawMessages(inner, defaultRole));
    return result;
  }
  if (typeof inner === 'string') {
    const innerParsed = parseJsonString(inner);
    if (innerParsed === undefined) {
      if (inner.trim()) {
        result.push({role: defaultRole, content: inner});
      }
      return result;
    }
    result.push(...rawMessagesFromValue(innerParsed, defaultRole));
  }
  return result;
}

/**
 * Legacy Vercel/AI SDK shape: `{system?, prompt}`.
 */
function unwrapSystemPrompt(value: UnknownRecord): RawMessage[] {
  const result: RawMessage[] = [];
  const system = value.system;
  if (system) {
    result.push({role: 'system', roleExplicit: true, content: system});
  }
  if (value.prompt) {
    result.push({role: 'user', roleExplicit: true, content: value.prompt});
  }
  return result;
}

function toRawMessage(item: unknown, defaultRole: string): RawMessage | null {
  if (typeof item === 'string') {
    return item.trim() ? {role: defaultRole, content: item} : null;
  }
  if (!isRecord(item)) {
    return null;
  }

  const role = getStringField(item, 'role');
  if (Array.isArray(item.parts)) {
    return {
      role: role ?? defaultRole,
      roleExplicit: role !== undefined,
      parts: item.parts,
    };
  }
  if (item.content !== undefined) {
    return {
      role: role ?? defaultRole,
      roleExplicit: role !== undefined,
      content: item.content,
    };
  }
  if (item.completion !== undefined) {
    return {
      role: role ?? defaultRole,
      roleExplicit: role !== undefined,
      content: item.completion,
    };
  }
  return null;
}

function resolveMessageContent(msg: RawMessage, role: string): unknown {
  if (msg.parts) {
    return collapseParts(msg.parts);
  }
  const parsed = parseJsonContentPreservingPrimitives(msg.content);
  return role === 'tool' ? parsed : renderTextContent(parsed);
}

function renderTextContent(content: unknown): unknown {
  return Array.isArray(content) ? extractTextFromContentParts(content) : content;
}

/**
 * Collapses a parts array into the single content value a message should
 * render. Prefers text (plus file redaction placeholders); falls back to
 * structured objects, then tool_calls, then tool_call_responses.
 */
function collapseParts(parts: unknown[]): unknown {
  const buckets = bucketParts(parts);

  if (buckets.hasRenderableTextPart) {
    return buckets.textParts.join('\n');
  }
  if (buckets.objectParts.length > 0) {
    return buckets.objectParts.length === 1
      ? buckets.objectParts[0]
      : buckets.objectParts;
  }
  if (buckets.toolCalls.length > 0) {
    return buckets.toolCalls;
  }
  if (buckets.toolResponses.length > 0) {
    return buckets.toolResponses.map(response => response.result).join('\n');
  }

  return undefined;
}

function bucketParts(parts: unknown[]): PartBuckets {
  const buckets: PartBuckets = {
    hasRenderableTextPart: false,
    textParts: [],
    objectParts: [],
    toolCalls: [],
    toolResponses: [],
  };
  for (const part of parts) {
    if (!isRecord(part)) {
      continue;
    }

    const partType = getPartType(part);
    if (isFileContentPartType(partType)) {
      buckets.hasRenderableTextPart = true;
      buckets.textParts.push(redactedFileContent(part));
      continue;
    }
    if (partType === 'text') {
      buckets.hasRenderableTextPart = true;
      const text = getTextPartContent(part, {trim: true});
      if (text) {
        buckets.textParts.push(text);
      }
      continue;
    }
    if (!partType) {
      const text = getTextPartContent(part, {trim: true});
      if (text) {
        buckets.textParts.push(text);
      }
      continue;
    }
    if (partType === 'object') {
      buckets.objectParts.push(part);
      continue;
    }
    if (partType === 'tool_call') {
      buckets.toolCalls.push(part);
      continue;
    }
    if (partType === 'tool_call_response') {
      buckets.toolResponses.push(part);
    }
  }

  return buckets;
}

function selectAssistantMessages(rawMessages: RawMessage[]): RawMessage[] {
  const hasRole = rawMessages.some(m => m.roleExplicit === true);
  if (hasRole) {
    return rawMessages.filter(m => m.role === 'assistant');
  }
  return [rawMessages[rawMessages.length - 1]!];
}

function appendOutputFromMessage(
  msg: RawMessage,
  buckets: {objectParts: unknown[]; textParts: string[]; toolCallParts: unknown[]}
): void {
  const {textParts, objectParts} = buckets;

  if (msg.parts) {
    appendOutputFromParts(msg.parts, buckets);
    return;
  }

  const content = parseJsonContentPreservingPrimitives(msg.content);
  if (content === undefined || content === null) {
    return;
  }
  if (typeof content === 'string' && content) {
    textParts.push(content);
  } else if (Array.isArray(content)) {
    const extracted = extractTextFromContentParts(content);
    if (extracted) {
      textParts.push(extracted);
    }
  } else if (content !== null && typeof content === 'object') {
    objectParts.push(content);
  }
}

function appendOutputFromParts(
  parts: unknown[],
  buckets: {objectParts: unknown[]; textParts: string[]; toolCallParts: unknown[]}
): void {
  const {textParts, toolCallParts, objectParts} = buckets;
  for (const part of parts) {
    const partType = getPartType(part);
    if (partType === 'text' && isRecord(part)) {
      const text = getStringField(part, 'content') ?? getStringField(part, 'text');
      if (text) {
        textParts.push(text);
      }
      continue;
    }
    if (partType === 'tool_call') {
      toolCallParts.push(part);
      continue;
    }
    if (partType === 'object') {
      objectParts.push(part);
      continue;
    }
    if (isFileContentPartType(partType) && isRecord(part)) {
      textParts.push(redactedFileContent(part));
    }
  }
}

const FILE_CONTENT_PARTS = ['blob', 'uri', 'file'] as const;
const FILE_CONTENT_PART_TYPES = new Set<string>(FILE_CONTENT_PARTS);
type FileContentPartType = (typeof FILE_CONTENT_PARTS)[number];
function parseJsonContentPreservingPrimitives(value: unknown): unknown {
  const parsed = tryParseJsonRecursive(value);
  if (
    typeof value === 'string' &&
    (parsed === null || typeof parsed === 'number' || typeof parsed === 'boolean')
  ) {
    return value;
  }
  return parsed;
}

function parseJsonString(value: string): unknown {
  try {
    return JSON.parse(value);
  } catch {
    return undefined;
  }
}

function looksLikeJson(raw: string): boolean {
  const trimmed = raw.trim();
  if (!trimmed) {
    return false;
  }
  const first = trimmed[0];
  return first === '[' || first === '{' || first === '"';
}

function extractTextFromContentParts(parts: unknown[]): string {
  return bucketParts(parts).textParts.join('\n');
}

function isRecord(value: unknown): value is UnknownRecord {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function getStringField(record: UnknownRecord, key: string): string | undefined {
  const value = record[key];
  return typeof value === 'string' && value.length > 0 ? value : undefined;
}

function getPartType(value: unknown): string | undefined {
  return isRecord(value) ? getStringField(value, 'type') : undefined;
}

function isFileContentPartType(value: unknown): value is FileContentPartType {
  return typeof value === 'string' && FILE_CONTENT_PART_TYPES.has(value);
}

function getMimeType(record: UnknownRecord): string {
  return getStringField(record, 'mime_type') ?? 'unknown';
}

function getTextPartContent(
  record: UnknownRecord,
  options: {trim?: boolean} = {}
): string | undefined {
  const text = getStringField(record, 'text') ?? getStringField(record, 'content');
  return options.trim ? text?.trim() : text;
}

function redactedFileContent(record: UnknownRecord): string {
  return `\n\n[redacted content of type "${getMimeType(record)}"]\n\n`;
}

File: src/sentry/utils/ai_message_normalizer.py
import json  # noqa: S003
from typing import Any, TypedDict

FILTERED = "[Filtered]"
FILE_CONTENT_PARTS = ("blob", "uri", "file")
_INVALID_JSON = object()


class AIMessage(TypedDict):
    role: str
    content: Any


class AIOutputResult(TypedDict):
    response_text: str | None
    response_object: str | None
    tool_calls: str | None


class RawMessage(TypedDict, total=False):
    role: str
    content: Any
    parts: list[Any]
    role_explicit: bool


class PartBuckets(TypedDict):
    has_renderable_text_part: bool
    text_parts: list[str]
    object_parts: list[Any]
    tool_calls: list[Any]
    tool_responses: list[dict[str, Any]]


# Keep this parser mirrored with
# static/app/views/insights/pages/agents/utils/aiMessageNormalizer.ts.
# AI SDKs emit inconsistent shapes and their specs keep changing, so update both
# parsers together whenever adding or changing a supported format.
def normalize_to_messages(raw: Any, default_role: str) -> list[AIMessage] | None:
    messages = _normalize_raw_messages(
        _raw_messages_from_attribute(raw, default_role), default_role
    )
    return messages or None


def extract_assistant_output(raw: Any, default_role: str) -> AIOutputResult:
    raw_messages = _raw_messages_from_attribute(raw, default_role)
    if not raw_messages:
        return _empty_output()

    selected = _select_assistant_messages(raw_messages)
    return _output_from_messages(selected)


def stringify_message_content(content: Any) -> str | None:
    if isinstance(content, str):
        return content or None

    try:
        return json.dumps(content)
    except TypeError:
        rendered = str(content)
        return rendered or None


def _empty_output() -> AIOutputResult:
    return {"response_text": None, "response_object": None, "tool_calls": None}


def _normalize_raw_messages(raw_messages: list[RawMessage], default_role: str) -> list[AIMessage]:
    normalized: list[AIMessage] = []
    for msg in raw_messages:
        role = msg.get("role") or default_role
        content = _resolve_message_content(msg, role)
        if content is None or content == "":
            continue
        normalized.append({"role": role, "content": content})
    return normalized


def _output_from_messages(messages: list[RawMessage]) -> AIOutputResult:
    text_parts: list[str] = []
    tool_call_parts: list[Any] = []
    object_parts: list[Any] = []
    for msg in messages:
        _append_output_from_message(
            msg,
            text_parts=text_parts,
            tool_call_parts=tool_call_parts,
            object_parts=object_parts,
        )

    response_object = None
    if object_parts:
        response_object = json.dumps(object_parts[0] if len(object_parts) == 1 else object_parts)

    return {
        "response_text": "\n".join(text_parts) if text_parts else None,
        "response_object": response_object,
        "tool_calls": json.dumps(tool_call_parts) if tool_call_parts else None,
    }


def _raw_messages_from_attribute(raw: Any, default_role: str) -> list[RawMessage]:
    parsed = _parse_attribute(raw)
    if parsed is _INVALID_JSON:
        return []
    return _raw_messages_from_value(parsed, default_role)


def _parse_attribute(raw: Any) -> Any:
    if not isinstance(raw, str):
        return raw
    if not _looks_like_json(raw):
        return raw
    return _parse_json_string(raw, fallback=_INVALID_JSON)


def _raw_messages_from_value(value: Any, default_role: str) -> list[RawMessage]:
    if isinstance(value, str):
        return [{"role": default_role, "content": value}] if value.strip() else []

    if isinstance(value, list):
        return _collect_raw_messages(value, default_role)

    if not isinstance(value, dict):
        return []

    if "messages" in value:
        return _unwrap_messages_field(value, default_role)

    if "prompt" in value:
        return _unwrap_system_prompt(value)

    single = _to_raw_message(value, default_role)
    return [single] if single else []


def _collect_raw_messages(items: list[Any], default_role: str) -> list[RawMessage]:
    out: list[RawMessage] = []
    for item in items:
        msg = _to_raw_message(item, default_role)
        if msg:
            out.append(msg)
    return out


def _unwrap_messages_field(value: dict[str, Any], default_role: str) -> list[RawMessage]:
    result: list[RawMessage] = []
    system = value.get("system")
    if system:
        result.append({"role": "system", "role_explicit": True, "content": system})

    inner = value.get("messages")
    if isinstance(inner, list):
        result.extend(_collect_raw_messages(inner, default_role))
        return result

    if isinstance(inner, str):
        parsed_inner = _parse_json_string(inner, fallback=_INVALID_JSON)
        if parsed_inner is _INVALID_JSON:
            if inner.strip():
                result.append({"role": default_role, "content": inner})
            return result
        result.extend(_raw_messages_from_value(parsed_inner, default_role))

    return result


def _unwrap_system_prompt(value: dict[str, Any]) -> list[RawMessage]:
    result: list[RawMessage] = []
    system = value.get("system")
    if system:
        result.append({"role": "system", "role_explicit": True, "content": system})
    prompt = value.get("prompt")
    if prompt:
        result.append({"role": "user", "role_explicit": True, "content": prompt})
    return result


def _to_raw_message(item: Any, default_role: str) -> RawMessage | None:
    if isinstance(item, str):
        return {"role": default_role, "content": item} if item.strip() else None

    if not isinstance(item, dict):
        return None

    raw_role = item.get("role")
    role = raw_role if isinstance(raw_role, str) and raw_role else None
    if isinstance(item.get("parts"), list):
        return {
            "role": role or default_role,
            "role_explicit": role is not None,
            "parts": item["parts"],
        }
    if "content" in item:
        return {
            "role": role or default_role,
            "role_explicit": role is not None,
            "content": item.get("content"),
        }
    if "completion" in item:
        return {
            "role": role or default_role,
            "role_explicit": role is not None,
            "content": item.get("completion"),
        }
    return None


def _resolve_message_content(msg: RawMessage, role: str) -> Any:
    if "parts" in msg:
        return _collapse_parts(msg["parts"])

    content = _try_parse_json_recursive(msg.get("content"))
    return content if role == "tool" else _render_text_content(content)


def _render_text_content(content: Any) -> Any:
    if isinstance(content, list):
        return _extract_text_from_content_parts(content)
    return content


def _try_parse_json_recursive(value: Any) -> Any:
    if not isinstance(value, str):
        return value

    parsed_value = _parse_json_string(value, fallback=_INVALID_JSON)
    if parsed_value is _INVALID_JSON:
        return value

    if parsed_value is None or isinstance(parsed_value, (bool, int, float)):
        return value
    if isinstance(parsed_value, list):
        return [_try_parse_json_recursive(item) for item in parsed_value]
    return parsed_value


def _collapse_parts(parts: list[Any]) -> Any:
    buckets = _bucket_parts(parts)

    if buckets["has_renderable_text_part"]:
        return "\n".join(buckets["text_parts"])
    if buckets["object_parts"]:
        return (
            buckets["object_parts"][0]
            if len(buckets["object_parts"]) == 1
            else buckets["object_parts"]
        )
    if buckets["tool_calls"]:
        return buckets["tool_calls"]
    if buckets["tool_responses"]:
        return "\n".join(str(response.get("result", "")) for response in buckets["tool_responses"])

    return None


def _bucket_parts(parts: list[Any]) -> PartBuckets:
    buckets: PartBuckets = {
        "has_renderable_text_part": False,
        "text_parts": [],
        "object_parts": [],
        "tool_calls": [],
        "tool_responses": [],
    }

    for part in parts:
        if not isinstance(part, dict):
            continue

        part_type = part.get("type")
        if part_type in FILE_CONTENT_PARTS:
            buckets["has_renderable_text_part"] = True
            buckets["text_parts"].append(_redacted_file_content(part))
            continue
        if part_type == "text":
            buckets["has_renderable_text_part"] = True
            text = _text_from_part(part, strip=True)
            if text:
                buckets["text_parts"].append(text)
            continue
        if not part_type:
            text = _text_from_part(part, strip=True)
            if text:
                buckets["text_parts"].append(text)
            continue
        if part_type == "object":
            buckets["object_parts"].append(part)
            continue
        if part_type == "tool_call":
            buckets["tool_calls"].append(part)
            continue
        if part_type == "tool_call_response":
            buckets["tool_responses"].append(part)

    return buckets


def _select_assistant_messages(raw_messages: list[RawMessage]) -> list[RawMessage]:
    has_role = any(msg.get("role_explicit") is True for msg in raw_messages)
    if has_role:
        return [msg for msg in raw_messages if msg.get("role") == "assistant"]
    return [raw_messages[-1]]


def _append_output_from_message(
    msg: RawMessage,
    *,
    text_parts: list[str],
    tool_call_parts: list[Any],
    object_parts: list[Any],
) -> None:
    if "parts" in msg:
        _append_output_from_parts(
            msg["parts"],
            text_parts=text_parts,
            tool_call_parts=tool_call_parts,
            object_parts=object_parts,
        )
        return

    content = _try_parse_json_recursive(msg.get("content"))
    if content is None:
        return
    if isinstance(content, str) and content:
        text_parts.append(content)
    elif isinstance(content, list):
        text = _extract_text_from_content_parts(content)
        if text:
            text_parts.append(text)
    elif isinstance(content, dict):
        object_parts.append(content)


def _append_output_from_parts(
    parts: list[Any],
    *,
    text_parts: list[str],
    tool_call_parts: list[Any],
    object_parts: list[Any],
) -> None:
    for part in parts:
        if not isinstance(part, dict):
            continue

        part_type = part.get("type")
        if part_type == "text":
            text = part.get("content") or part.get("text")
            if isinstance(text, str) and text:
                text_parts.append(text)
            continue
        if part_type == "tool_call":
            tool_call_parts.append(part)
            continue
        if part_type == "object":
            object_parts.append(part)
            continue
        if part_type in FILE_CONTENT_PARTS:
            text_parts.append(_redacted_file_content(part))


def _parse_json_string(value: str, *, fallback: Any) -> Any:
    try:
        return json.loads(value)
    except (json.JSONDecodeError, TypeError):
        return fallback


def _looks_like_json(raw: str) -> bool:
    stripped = raw.strip()
    if not stripped:
        return False
    return stripped[0] in ("[", "{", '"')


def _extract_text_from_content_parts(parts: list[Any]) -> str:
    return "\n".join(_bucket_parts(parts)["text_parts"])


def _text_from_part(part: dict[str, Any], *, strip: bool = False) -> str | None:
    text = part.get("text") or part.get("content")
    if not isinstance(text, str) or not text:
        return None
    return text.strip() if strip else text


def _redacted_file_content(part: dict[str, Any]) -> str:
    return f'\n\n[redacted content of type "{part.get("mime_type") or "unknown"}"]\n\n'
